tests/training/trainer_test.py

```diff
@@ -89,3 +109,113 @@ def forward(self, **kwargs):  # pylint: disable=arguments-differ,unused-argument
+        #   Check the resulting checkpoints.  Should then have models at epochs
+        #       2, 4, plus the last two at 5 and 6.
+        class WaitingIterator(BasicIterator):
+            # pylint: disable=arguments-differ
+            def _create_batches(self, *args, **kwargs):
+                time.sleep(0.5)
+                return super(WaitingIterator, self)._create_batches(*args, **kwargs)
+
+        iterator = WaitingIterator(batch_size=2)
+        iterator.index_with(self.vocab)
+
+        trainer = Trainer(self.model, self.optimizer,
+                          iterator, self.instances, num_epochs=6,
+                          serialization_dir=self.TEST_DIR,
+                          num_serialized_models_to_keep=2,
+                          keep_serialized_model_every_num_seconds=1)
+        trainer.train()
+
+        # Now check the serialized files
+        for prefix in ['model_state_epoch_*', 'training_state_epoch_*']:
+            file_names = glob.glob(os.path.join(self.TEST_DIR, prefix))
+            epochs = [int(re.search(r"_([0-9])\.th", fname).group(1))
+                      for fname in file_names]
+            # epoch N has N-1 in file name
+            assert sorted(epochs) == [1, 3, 4, 5]
+
+    def test_trainer_saves_models_at_specified_interval(self):
+        trainer = Trainer(self.model, self.optimizer,
+                          self.iterator, self.instances, num_epochs=2,
+                          serialization_dir=self.TEST_DIR,
+                          model_save_interval=0.0001)
+        trainer.train()
+
+        # Now check the serialized files for models saved during the epoch.
+        prefix = 'model_state_epoch_*'
+        file_names = sorted(glob.glob(os.path.join(self.TEST_DIR, prefix)))
+        epochs = [re.search(r"_([0-9\.]+)\.th", fname).group(1)
+                  for fname in file_names]
+        # We should have checkpoints at the end of each epoch and during each, e.g.
+        # [0.timestamp, 0, 1.timestamp, 1]
+        assert len(epochs) == 4
+        assert epochs[3] == '1'
+        assert '.' in epochs[0]
+
+        # Now make certain we can restore from timestamped checkpoint.
+        # To do so, remove the checkpoint from the end of epoch 1&2, so
+        # that we are forced to restore from the timestamped checkpoints.
+        for k in range(2):
+            os.remove(os.path.join(self.TEST_DIR, 'model_state_epoch_{}.th'.format(k)))
+            os.remove(os.path.join(self.TEST_DIR, 'training_state_epoch_{}.th'.format(k)))
+        os.remove(os.path.join(self.TEST_DIR, 'best.th'))
+
+        restore_trainer = Trainer(self.model, self.optimizer,
+                                  self.iterator, self.instances, num_epochs=2,
+                                  serialization_dir=self.TEST_DIR,
+                                  model_save_interval=0.0001)
+        epoch, _ = restore_trainer._restore_checkpoint() # pylint: disable=protected-access
+        assert epoch == 2
```

Comment:
Can you check that batch_num_total is the correct value after restoring, because if it isn't, tensorboard logging will very wrong.

Reply:
done